// @vue-macros/reactivity-transform/dist/chunk-TJ26NVC5.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});





var _chunk4PJYZQLLjs = require('./chunk-4PJYZQLL.js');

// src/index.ts
var _unplugin = require('unplugin');









var _common = require('@vue-macros/common');
function resolveOption(options) {
  const version = options.version || _common.detectVueVersion.call(void 0, );
  return {
    include: [_common.REGEX_SRC_FILE, _common.REGEX_VUE_SFC, _common.REGEX_SETUP_SFC, _common.REGEX_VUE_SUB],
    exclude: [_common.REGEX_NODE_MODULES],
    ...options,
    version
  };
}
var name = "unplugin-vue-reactivity-transform";
var src_default = _unplugin.createUnplugin.call(void 0, 
  (userOptions = {}) => {
    const options = resolveOption(userOptions);
    const filter = _common.createFilter.call(void 0, options);
    return {
      name,
      enforce: "pre",
      resolveId(id) {
        if (id === _common.normalizePath.call(void 0, _chunk4PJYZQLLjs.helperId))
          return id;
      },
      loadInclude(id) {
        return _common.normalizePath.call(void 0, id) === _chunk4PJYZQLLjs.helperId;
      },
      load(id) {
        if (_common.normalizePath.call(void 0, id) === _chunk4PJYZQLLjs.helperId)
          return _chunk4PJYZQLLjs.code_default;
      },
      transformInclude(id) {
        return filter(id);
      },
      transform(code, id) {
        if (_common.REGEX_VUE_SFC.test(id) || _common.REGEX_SETUP_SFC.test(id) || _common.REGEX_VUE_SUB.test(id)) {
          return _chunk4PJYZQLLjs.transformVueSFC.call(void 0, code, id);
        } else if (_chunk4PJYZQLLjs.shouldTransform.call(void 0, code)) {
          return _chunk4PJYZQLLjs.transform.call(void 0, code, {
            filename: id,
            sourceMap: true
          });
        }
      }
    };
  }
);



exports.src_default = src_default;
