// @vue-macros/reactivity-transform/dist/chunk-TEMCFQ2G.mjs
import {
  code_default,
  helperId,
  shouldTransform,
  transform,
  transformVueSFC
} from "./chunk-OZECVUFY.mjs";

// src/index.ts
import { createUnplugin } from "unplugin";
import {
  REGEX_NODE_MODULES,
  REGEX_SETUP_SFC,
  REGEX_SRC_FILE,
  REGEX_VUE_SFC,
  REGEX_VUE_SUB,
  createFilter,
  detectVueVersion,
  normalizePath
} from "@vue-macros/common";
function resolveOption(options) {
  const version = options.version || detectVueVersion();
  return {
    include: [REGEX_SRC_FILE, REGEX_VUE_SFC, REGEX_SETUP_SFC, REGEX_VUE_SUB],
    exclude: [REGEX_NODE_MODULES],
    ...options,
    version
  };
}
var name = "unplugin-vue-reactivity-transform";
var src_default = createUnplugin(
  (userOptions = {}) => {
    const options = resolveOption(userOptions);
    const filter = createFilter(options);
    return {
      name,
      enforce: "pre",
      resolveId(id) {
        if (id === normalizePath(helperId))
          return id;
      },
      loadInclude(id) {
        return normalizePath(id) === helperId;
      },
      load(id) {
        if (normalizePath(id) === helperId)
          return code_default;
      },
      transformInclude(id) {
        return filter(id);
      },
      transform(code, id) {
        if (REGEX_VUE_SFC.test(id) || REGEX_SETUP_SFC.test(id) || REGEX_VUE_SUB.test(id)) {
          return transformVueSFC(code, id);
        } else if (shouldTransform(code)) {
          return transform(code, {
            filename: id,
            sourceMap: true
          });
        }
      }
    };
  }
);

export {
  src_default
};
